### ISSUE TITLE
Esc does not close flameshot --gui

### ISSUE THREAD

**Issue by mmahmoudian** [flameshot-org/flameshot#1072]
**Flameshot version**
Flameshot v0.8.5 (AppImage)
Compiled with Qt 5.12.8

**Describe the bug**
When any application tries to take over the screen, there will be _almost_ no way to close Flameshot (other than pressing the X button). This happened to me during the following scenarios:
1. I was in the waiting room of a zoom call and I was taking a screenshot with Flameshot that they let me in the Zoom call and then there was no way to close Flameshot (<kbd>Esc</kbd> was not working)
2. While taking a screenshot I pressed <kbd>Alt</kbd><kbd>Tab</kbd> accidentally and then the <kbd>Esc</kbd> was non functional. (This scenario was much easier to repeat)

**To Reproduce**
1. `flameshot --gui`
2. press <kbd>Alt</kbd><kbd>Tab</kbd> (or what ever shortcut you use to switch  windows
3. press <kbd>Esc</kbd> to close

**Expected behavior**
<kbd>Esc</kbd> works as long as `flameshot --gui` is open.

**System Information**
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.5 LTS (Ubuntu Budgie)
DE: Budgie
Release:	18.04
Codename:	bionic
Display server protocol: X11

**Comment by mmahmoudian** [comment]
@ZetaoYang Are you sure it does not behave in the same way as in Windows? Because the <kbd>Alt</kbd><kbd>Tab</kbd> works similarly in Windows.

**Comment by ZetaoYang** [comment]
@mmahmoudian CLI option`flameshot gui` only works on Linux.

**Comment by borgmanJeremy** [comment]
You can trigger the bug without using the CLI. If you start taking a screenshot from the tray icon and alt tab it will appear the system is frozen and you will need to drop to tty to kill flameshot.

I think the fix is to minimize or close an active edit if flameshot looses screen focus.

**Comment by ZetaoYang** [comment]
> You can trigger the bug without using the CLI. If you start taking a screenshot from the tray icon and alt tab it will appear the system is frozen and you will need to drop to tty to kill flameshot.
> 
> I think the fix is to minimize or close an active edit if flameshot looses screen focus.

On Windows, this bug will not appear.

**Comment by mmahmoudian** [comment]
@ZetaoYang 
> @mmahmoudian CLI option`flameshot gui` only works on Linux currently.

Yes, I didn't know how to refer to the UI that opens when user wants to take screenshot, so I thought the most accurate way would be to write the command to clarify what do I mean.

I think we should come up with some sort of uniform terminology for different things in Flameshot to avoid confusion.

-------
@borgmanJeremy 
> I think the fix is to minimize or close an active edit if flameshot looses screen focus.

I think minimizing is a good option as it gives the user the freedom to return to the screenshot process and do the annotations they need.

**Comment by avinashrocks1990** [comment]
Flameshot GUI won't close on pressing escape.
To Reproduce

    Press F12 to Launch Yakuake (drop down terminal)
    flameshot gui
    press Esc to close

Expected behavior
Esc works as long as flameshot gui is open.

System Information
Distributor ID: Manjaro
Description: Manjaro Unstable
DE: KDE Plasma
Release: 5.20.4


Display server protocol: X11

**Comment by mmahmoudian** [comment]
@avinashrocks1990 Please also report/confirm the following:
1. your Flameshot version
2. The way you have installed Flameshot (AppImage, AUR, pacman, snap, compile from source, ...)

**Comment by avinashrocks1990** [comment]
Flameshot 0.8.5-1
I have installed it using pacman(official repo), i have also tried flameshot-git(AUR).

**Comment by davama** [comment]
Just to add to this post:

```
flameshot -v
Flameshot v0.9.0
Compiled with Qt 5.15.2
```

When I run from the CLI `flameshot gui` I cannot exit with `ESC`
Only until I create a box screenshot with my mouse, then I can press the `x` button to kill `flameshot`

Using xmonad as my WM if that matters.

**Comment by danyeet** [comment]
I am using xmonad too, and I have encountered this issue.

**Comment by 3vi1** [comment]
I've been running into this a long time myself, using Plasma desktop.  Here's how I can recreate it at will:

Good Case:
- Press PrtSc... flameshot opens... press Esc.... flameshot closes as expected.

Bad Case:
- Press PrtSc... flameshot opens... press F12 (which is bound to yakuake)... press Esc... flameshot stays open.  Press F12 again (which closes yakuake in the unseeable background).... press Esc.... flameshot still doesn't close (I think because it's lost focus).

**Comment by CatalinPuscoci** [comment]
The overlay loses focus on activation in sway too. Hotekeys and esc do not work.

**Comment by maguilara** [comment]
I'm facing the same issue myself, but I think mine was working until a few months ago (I just came back to my arch partition). Today I tested a new key combo, and then I reverted to the same key (PrntScr), and I think just then it started to fail. Not really sure if after setting the new key combo it started to fail since I haven't taking too many screenshots recently but it feels that way. I use the KDE shortcut with the action flameshot gui

Operating System: Arch Linux
KDE Plasma Version: 5.22.5
KDE Frameworks Version: 5.85.0
Qt Version: 5.15.2
Kernel Version: 5.10.63-1-lts (64-bit)
Graphics Platform: X11
Flameshot: 0.10.1-1

**Comment by mmahmoudian** [comment]
As pointed out by @adamant-pwn in #1924, if Flameshot is invoked while another application is loading (e.g Steam), that other application will steal the focus from Flameshot and consequently Flameshot appears to unresponsive from user's perspective simply because all the keyboard input is directed to the other app.

@borgmanJeremy & @veracioux is it possible to detect if Flameshot loses focus so that we can ask the OS to get the focus back to Flameshot (Perhaps `QWidget::activateWindow()` or `QWidget::setFocus()` or `QWidget::raise()`)?

**Comment by borgmanJeremy** [comment]
Yeah there is a Qt signal for that: https://doc.qt.io/qt-5/qapplication.html#focusChanged

**Comment by veracioux** [comment]
@mmahmoudian @borgmanJeremy What if the user wants to remove focus from flameshot intentionally? What if flameshot bugs out on them for some reason and then they can't close it because they can't focus any other program? My suggestion is to do the following if flameshot loses focus: hide the capture gui and open a small dialog window with a button that says "Resume capture" or something.

**Comment by mmahmoudian** [comment]
@veracioux that's a good point (although I don't see why anyone want to switch applications apart from a terminal or task manager to kill Flameshot if it somehow crashes). Perhaps it is a good idea to explain to user a bit more on why this has happened and even what has stole the focus. This way we will see less bug reports like "Flameshot stops for no reason and asks to be resumed".

**Comment by jfly** [comment]
As a reminder, disabling `Qt::BypassWindowManagerHint` makes this way less confusing. From https://github.com/flameshot-org/flameshot/pull/731#issuecomment-719767364:

>   - It also makes it possible for me to clearly see when flameshot has focus, and give it focus, which I'd argue is a workaround/fix for https://github.com/flameshot-org/flameshot/issues/1072, and also has some nice properties as @filipkilibarda mentioned: "Allows the user to switch workspaces while in the screenshot GUI..."

Unfortunately, we had to revert my attempt to make that change. See https://github.com/flameshot-org/flameshot/issues/784#issuecomment-920244344 for details.

**Comment by dorian-marchal** [comment]
> although I don't see why anyone want to switch applications apart from a terminal or task manager to kill Flameshot if it somehow crashes

FYI I encountered this bug while trying to open an app after taking a screenshot with:

```
flameshot gui && xdg-open ~/Pictures
```
*Note : the above command doesn't work as expected, cf. [this unrelated issue](https://github.com/flameshot-org/flameshot/issues/1935) I just opened.*

It took me some time to figure out because I didn't know Flameshot covered the whole screen with a still image of the desktop at the time of the capture, so I didn't see the focus-stealing app appear.

**Comment by johnrichardrinehart** [comment]
@veracioux It seems to improve this a bit for me. However, I can only kill by <kbd>Esc</kbd> after selecting a region. <kbd>Esc</kbd> is a no-op in the below state. Is this expected?

![image](https://user-images.githubusercontent.com/6321578/137123309-da8199f3-5491-4bee-bf43-0459e305485d.png)

**Comment by johnrichardrinehart** [comment]
That was a good idea. I tried to use my "application cycle" keyboard shortcut. No luck. The changes in `master` are definitely an improvement. Any timeline for tagging a new release? I want to bump the version in the package manager (`NixOS`).

**Comment by dginovker** [comment]
Fresh install of KDE Arch Linux has the issue, existing KDE Manjaro install doesn't -

Both have Flameshot v0.10.2, compiled with Qt 5.15.2

Happens regardless of activating `flameshot gui` via shortcut or command line.

Other Arch Linux specs:
* x86_64
* Linux zen kernel
* Plasma 5.23.4, KWin

**Comment by mmahmoudian** [comment]
> Flameshot --version though shows v0.10.1, so flameshot-git might be outdated?

No it is not outdated. It is actually up to date. The reason behind having v0.10.1 instead of v0.10.2 is that the v0.10.2 was a cherrypicked release just for fix some stuff for Gnome 41 users. The AUR is what will eventually called v0.11.0 with few more things merged into it.

**Comment by johnrichardrinehart** [comment]
@dginovker #2159 and #2161 have been opened to address this version inconsistency (tagged release is `v0.10.2`, but source code hasn't been updated correspondingly so `--version` still outputs `v0.10.1`).

**Comment by shbitsus** [comment]
version: 0.9.0+ds1-2+deb11u2  XFCE.
I just experienced same, or very similar:  I was annotating a screen shot when I pressed some unknown key resulting in flameshot selection changed to include all three of my displays with the Line tool selected.  I could draw lines anywhere but ESC does nothing. Tried CTRL+ALT+F2, X, ESC, nothing would give me back control.  I ended up using a second computer to ssh in and kill flameshot.  After reading this issue, I tested doing ALT+Tab during an edit.  It didn't change my selection size, but it did get similarly stuck:  unable to escape flameshot edit mode, able to draw anywhere on my screens, even outside my selection box, and no way to leave or otherwise interact with my system.  I tested several more times.  If the flameshot GUI 'X' button is accessible, clicking it still works.

**Comment by digitalstudium** [comment]
$ flameshot -v
Flameshot v12.1.0 (Debian 12.1.0-2)
Compiled with Qt 5.15.8

Same problem.
